### PR TITLE
Add min and max facet stats

### LINKED ISSUE #3426: Return stats for numerical facets like `min`, `max`
Related product discussion: https://github.com/meilisearch/product/discussions/117
Related specification: WIP
Also related to: https://github.com/meilisearch/instant-meilisearch/issues/926

### TODO

- [ ] For implementation, check with the product squad to get how it should be implemented
    - A prototype PR with only `min` and `max` is available: #3423. Do we need `avg` and `sum`?
- [ ] Merge the changes into main
- [ ] Update the specification

### Impacted teams

@meilisearch/integration-team: usable for https://github.com/meilisearch/instant-meilisearch/issues/926
@meilisearch/docs-team

### PR DESCRIPTION
**Pull request by dureuill**
# Pull Request

## Related issue
Fixes #3426

### User standpoint

- When using a `facets` parameter in search, the facets that have numeric values are displayed in a new section of the response called `facetStats` that contains, per facet, the numeric min and max value of the hits returned by the search.

<details>
<summary>
Sample request/response
</summary>

```json
❯ curl \
  -X POST 'http://localhost:7700/indexes/meteorites/search?facets=mass' \
  -H 'Content-Type: application/json' \
  --data-binary '{ "q": "LL6", "facets":["mass", "recclass"], "limit": 5 }' | jsonxf
{
  "hits": [
    {
      "name": "Niger (LL6)",
      "id": "16975",
      "nametype": "Valid",
      "recclass": "LL6",
      "mass": 3.3,
      "fall": "Fell"
    },
    {
      "name": "Appley Bridge",
      "id": "2318",
      "nametype": "Valid",
      "recclass": "LL6",
      "mass": 15000,
      "fall": "Fell",
      "_geo": {
        "lat": 53.58333,
        "lng": -2.71667
      }
    },
    {
      "name": "Athens",
      "id": "4885",
      "nametype": "Valid",
      "recclass": "LL6",
      "mass": 265,
      "fall": "Fell",
      "_geo": {
        "lat": 34.75,
        "lng": -87.0
      }
    },
    {
      "name": "Bandong",
      "id": "4935",
      "nametype": "Valid",
      "recclass": "LL6",
      "mass": 11500,
      "fall": "Fell",
      "_geo": {
        "lat": -6.91667,
        "lng": 107.6
      }
    },
    {
      "name": "Benguerir",
      "id": "30443",
      "nametype": "Valid",
      "recclass": "LL6",
      "mass": 25000,
      "fall": "Fell",
      "_geo": {
        "lat": 32.25,
        "lng": -8.15
      }
    }
  ],
  "query": "LL6",
  "processingTimeMs": 15,
  "limit": 5,
  "offset": 0,
  "estimatedTotalHits": 42,
  "facetDistribution": {
    "mass": {
      "110000": 1,
      "11500": 1,
      "1161": 1,
      "12000": 1,
      "1215.5": 1,
      "127000": 1,
      "15000": 1,
      "1676": 1,
      "1700": 1,
      "1710.5": 1,
      "18000": 1,
      "19000": 1,
      "220000": 1,
      "2220": 1,
      "22300": 1,
      "25000": 2,
      "265": 1,
      "271000": 1,
      "2840": 1,
      "3.3": 1,
      "3000": 1,
      "303": 1,
      "32000": 1,
      "34000": 1,
      "36.1": 1,
      "45000": 1,
      "460": 1,
      "478": 1,
      "483": 1,
      "5500": 2,
      "600": 1,
      "6000": 1,
      "67.8": 1,
      "678": 1,
      "680.5": 1,
      "6930": 1,
      "8": 1,
      "8300": 1,
      "840": 1,
      "8400": 1
    },
    "recclass": {
      "L/LL6": 3,
      "LL6": 39
    }
  },
  "facetStats": {
    "mass": {
      "min": 3.3,
      "max": 271000.0
    }
  }
}
```

</details>

## PR checklist
Please check if your PR fulfills the following requirements:
- [ ] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [ ] Have you read the contributing guidelines?
- [ ] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

File: milli/src/search/criteria/asc_desc.rs
Comment (Kerollmops) [inline milli/src/search/criteria/asc_desc.rs:210]: Not sure we want to use this syntax right now. But don't worry we will one day :)

**Review by Kerollmops — CHANGES_REQUESTED** [review]

File: milli/src/search/criteria/asc_desc.rs
Comment (dureuill) [reply milli/src/search/criteria/asc_desc.rs:210]: Awwwww, that's too bad, I like it 💔 

Any ETA as to when we want to use this? After rustfmt support lands https://github.com/rust-lang/rustfmt/issues/4914 ? Some other thing?

I guess the alternative is the good old:

```rust
let value = if let Ok(value) = std::str::from_utf8(value) { value } else { return Ok(None) };
```

Comment (dureuill) [reply milli/src/search/criteria/asc_desc.rs:210]: (also, I may or may not have smuggled a few `let-else`s in other PRs already 🙊)

Comment (Kerollmops) [reply milli/src/search/criteria/asc_desc.rs:210]: I would say waiting for the rustfmt issue to be closed and most of the common text editor syntax highlighters to be ready, so something like 6 months or so? What do you think? Is it too much?

Maybe we can plan to do a big pass one day to change all of the `if let else` in to `let else` statements?

Comment (dureuill) [reply milli/src/search/criteria/asc_desc.rs:210]: > I would say waiting for the rustfmt issue to be closed and most of the common text editor syntax highlighters to be ready, so something like 6 months or so? What do you think? Is it too much?

Ah, I was not aware that the syntax highlighters weren't ready (it works for me with vscode and r-a).

OK to wait for rusfmt support for now (although I'm a bit unsure if it will take more or less than six months)..., as we have an "automatic" way to desugar the `let else` to `if let else` at the moment (which is the way we were using up to now).

> Maybe we can plan to do a big pass one day to change all of the if let else in to let else statements?

I'd love for a `cargo clippy --fix` for this one. Maybe it's time I take a look at clippy 🤓

Comment (dureuill) [reply milli/src/search/criteria/asc_desc.rs]: removed 🙂

**Comment by dureuill** [comment]
@loiclec I implemented the changes we discussed this morning regarding the Ascending/DescendingFacetSorts. I'm taking the liberty to add you as a reviewer so you can take a look as these changes :pray:

**Comment by dureuill** [comment]
rebased on main after #3461

**Review by Kerollmops — APPROVED** [review]
Thank you very much for your work. It's pretty good. Nothing more to say than: "please, use `match`es instead of `if else` 😂".
You can merge it, I don't know if you want to wait for @loiclec's review.

File: milli/src/search/criteria/asc_desc.rs
Comment (loiclec) [inline milli/src/search/criteria/asc_desc.rs:259]: 👍 
However, I don't really understand why we use an `impl Iterator` instead of `Box<dyn Iterator` now. Was that necessary in order to put a pointer to the DB key inside?

File: milli/src/search/facet/facet_distribution.rs
Comment (loiclec) [inline milli/src/search/facet/facet_distribution.rs:311]: Is is possible for this condition to be false?
(Even if it is not, I am okay with keeping it, out of an excess of caution :) )

**Review by loiclec — APPROVED** [review]
Thanks Louis :) I couldn't spot anything wrong with it.

File: milli/src/search/criteria/asc_desc.rs
Comment (dureuill) [reply milli/src/search/criteria/asc_desc.rs:259]: I'm honestly not sure anymore if it was necesssary, I believe I just changed it as part of my refactoring. I like to avoid allocations when practical, but that line is not always easy to find. I don't mind changing it back if you prefer.

**Comment by dureuill** [comment]
thanks for the reviews both of you ❤️ 

bors merge

**Pull request merged**
